modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -854,4 +891,66 @@ static boolean isRecoverable(Throwable throwable) {
             return false;
         }
     }
+
+    /**
+     * Handles messages of {@link TxMessageGroup}.
+     *
+     * @see NetworkMessageHandler#onReceived(NetworkMessage, String, Long)
+     */
+    private void onReceiveTxNetworkMessage(NetworkMessage message, String senderConsistentId, @Nullable Long correlationId) {
+        if (!(message instanceof RwTransactionsFinishedRequest)) {
+            return;
+        }
+
+        assert correlationId != null : senderConsistentId;
+
+        int targetCatalogVersion = ((RwTransactionsFinishedRequest) message).targetSchemaVersion();
+
+        boolean finished = isRwTransactionsFinished(targetCatalogVersion);
+
+        clusterService.messagingService().respond(senderConsistentId, toRwTransactionsFinishedResponse(finished), correlationId);
+    }
+
+    /**
+     * Returns the transaction catalog version.
+     *
+     * @param beginTs Transaction begin timestamp.
+     */
+    private int txCatalogVersion(HybridTimestamp beginTs) {
+        return catalogService.activeCatalogVersion(beginTs.longValue());
+    }
+
+    /**
+     * Returns the transaction catalog version.
+     *
+     * @param txId Transaction ID.
+     */
+    private int txCatalogVersion(UUID txId) {
+        return txCatalogVersion(TransactionIds.beginTimestamp(txId));
+    }
+
+    /**
+     * Returns {@code true} iff the requested catalog version is registered locally and all RW transactions before it have completed,
+     * otherwise {@code false}.
+     *
+     * @param catalogVersion Catalog version to check.
+     */
+    boolean isRwTransactionsFinished(int catalogVersion) {
+        if (catalogVersion > catalogService.latestCatalogVersion()) {
```

Comment:
Here, we need to look at the 'currently active' version of the catalog (that is, the version that is active at ts=now), not just the 'latest known version of the catalog'

Reply:
At the moment, the latest version of the catalog is the last active version of the catalog, you can check it in the code.

Reply:
We discussed it in person and I corrected it.